src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+          TypeMismatchError : -2,
+          IndexSizeError : -2,
+          // LIBUSB_ERROR_ACCESS
+          SecurityError : -3,
+          …
+        };
+        console.error(error);
+        let errorCode = -99; // LIBUSB_ERROR_OTHER
+        if (error instanceof DOMException)
+        {
+          errorCode = ERROR_CODES[error.name] ?? errorCode;
+        }
+        else if (error instanceof RangeError || error instanceof TypeError)
+        {
+          errorCode = -2; // LIBUSB_ERROR_INVALID_PARAM
+        }
+        return {error: errorCode, value: undefined};
+      }
+    );
+    return Emval.toHandle(promise);
+  });
+// clang-format on
+
+val em_promise_catch(val &&promise) {
+  EM_VAL handle = promise.as_handle();
+  handle = em_promise_catch_impl(handle);
+  return val::take_ownership(handle);
+}
+
+// C++ struct representation for {value, error} object from above
+// (performs conversion in the constructor).
+struct promise_result {
+  libusb_error error;
+  val value;
+
+  promise_result(val &&result)
+      : error(static_cast<libusb_error>(result["error"].as<int>())),
+        value(result["value"]) {}
+
+  // C++ counterpart of the promise helper above that takes a promise, catches
+  // its error, converts to a libusb status and returns the whole thing as
+  // `promise_result` struct for easier handling.
+  static promise_result await(val &&promise) {
+    promise = em_promise_catch(std::move(promise));
+    return {promise.await()};
+  }
+};
+```
+
+Now I could use `promise_result::await` on any `Promise` returned from WebUSB operations and inspect its `error` and `value` fields separately.
+
+For example, retrieving a `val` representing a [`USBDevice`](https://developer.mozilla.org/docs/Web/API/USBDevice) from `libusb_device_handle`, calling its [`.open()`](https://developer.mozilla.org/docs/Web/API/USBDevice/open) method, awaiting its result and returning an error code as a libusb status code looks like this:
+
+```cpp
+int em_open(libusb_device_handle *handle) {
+  auto web_usb_device = get_web_usb_device(handle->dev);
+  return promise_result::await(web_usb_device.call<val>("open")).error;
+}
+```
+
+### Device enumeration
+
+Of course, before any device can be opened, libusb needs to retrieve a list of available devices. The backend must implement this operation via a `get_device_list` handler.
+
+The difficulty is that, unlike on other platforms, there is no way to enumerate all the connected USB devices on the web for security reasons. Instead, the flow is split into two parts. First, the web application can request devices with specific properties via [`navigator.usb.requestDevice()`](https://developer.mozilla.org/docs/Web/API/USB/requestDevice) and the user will manually choose which device they want to expose or they can reject the permission prompt. Then, you can list the already approved and connected devices via [`navigator.usb.getDevices()`](https://developer.mozilla.org/docs/Web/API/USB/getDevices).
```

Comment:
```suggestion
The difficulty is that, unlike on other platforms, there is no way to enumerate all the connected USB devices on the web for security reasons. Instead, the flow is split into two parts. First, the web application requests devices with specific properties via [`navigator.usb.requestDevice()`](https://developer.mozilla.org/docs/Web/API/USB/requestDevice) and the user will manually choose which device they want to expose or they can reject the permission prompt. Then, you can list the already approved and connected devices via [`navigator.usb.getDevices()`](https://developer.mozilla.org/docs/Web/API/USB/getDevices).
```

Reply:
Should the 2nd part of that sentence be changed to use present tense too perhaps? Something like:

```suggestion
The difficulty is that, unlike on other platforms, there is no way to enumerate all the connected USB devices on the web for security reasons. Instead, the flow is split into two parts. First, the web application requests devices with specific properties via [`navigator.usb.requestDevice()`](https://developer.mozilla.org/docs/Web/API/USB/requestDevice) and the user manually chooses which device they want to expose or rejects the permission prompt. Afterwards, the application can list the already approved and connected devices via [`navigator.usb.getDevices()`](https://developer.mozilla.org/docs/Web/API/USB/getDevices).
```

Reply:
Yes! Good catch.

Reply:
I updated the suggestion.